### PR TITLE
[NN] Add SparseBiasedMultiheadAttention

### LINKED ISSUE #4882: Add biased multi-head attention module
## 🔨Work Item

Add biased multi-head attention module for attention modification in graph transformer layers. Provide two separate APIs:
- [x] Dense weighted attention computation
- [ ] Sparse attention computation

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by rudongyu**
## Description

close #4882 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 1]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
 
-        attn = self.out_proj(attn.reshape(N, bsz, self.feat_size).transpose(0, 1))
+        attn = self.out_proj(
+            attn.reshape(N, bsz, self.feat_size).transpose(0, 1)
+        )
 
         return attn
+
+
+class SparseBiasedMultiheadAttention(nn.Module):
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:250]: The name becomes sooooo long, not sure if there is a more concise alternative.

Comment (rudongyu) [reply python/dgl/nn/pytorch/graph_transformer.py:250]: How about replace `MultiheadAttention` in all module names with `MHA`?

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 2]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+    r"""Sparse Multi-Head Attention Module with Graph Attention Bias.
+
+    Implement sparse attention computation restricted to neighborhoods of
+    nodes, as introduced in `A Generalization of Transformer Networks to
+    Graphs <https://arxiv.org/pdf/2012.09699>`__
```
[the remark this answers is not in the document]
Comment (rudongyu) [reply python/dgl/nn/pytorch/graph_transformer.py:255]: For their edge-feature-excluded design, attention is not modified with biases, so we allow the attn_bias to be `None` in forward computation. However, we leave this param for potentially integrating edge features encoding or other types of bias terms.

[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/nn/pytorch/graph_transformer.py:255]: It's better if we can reference a paper that actually uses biased attention over graph structure.

File: python/dgl/nn/pytorch/graph_transformer.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/graph_transformer.py:250]: Sounds good to me, though you will need to update the other modules accordingly for consistency

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 3]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+
+    .. math::
+
+        \text{Attn}_{ij}=\text{softmax}_{j\in\mathcal{N}(i)}(\dfrac{Q_iK_j^T}
+        {\sqrt{d}} \circ b)
+
+    :math:`Q_i` and :math:`K_j` are feature representation of nodes. :math:`d`
+    is the corresponding :attr:`feat_size`. :math:`b` is attention bias, which
+    can be additive or multiplicative according to the operator :math:`\circ`.
+    :math:`\mathcal{N}(i)` is the set of source nodes that have an edge to the
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:265]:
```suggestion
    :math:`\mathcal{N}(i)` is the set of nodes that have an edge to the
```

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 4]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+    destination node :math:`i`.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:266]:
```suggestion
    node :math:`i`.
```

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 5]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+
+    Parameters
+    ----------
+    feat_size : int
+        Feature size.
+    num_heads : int
+        Number of attention heads, by which attr:`feat_size` is divisible.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:273]:
```suggestion
        Number of attention heads, by which :attr:`feat_size` is divisible.
```

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 6]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+    bias : bool, optional
+        If True, it uses bias for linear projection. Default: True.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:275]: When may people want to disable it?

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 7]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+    attn_bias_type : str, optional
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:276]: This term does not come into effect now.

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 8]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        The type of attention bias used for modifying attention. Selected from
+        'add' or 'mul'. Default: 'add'.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:278]: Maybe document an option of `None` and `attn_bias` can only be `None` in this case.

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 9]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+
+        * 'add' is for additive attention bias.
+        * 'mul' is for multiplicative attention bias.
+    attn_drop : float, optional
+        Dropout probability on attention weights. Defalt: 0.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:283]:
```suggestion
        Dropout probability on attention weights. Default: 0.
```

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 10]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+
+    Examples
+    --------
+    >>> import dgl
+    >>> import torch as th
+    >>> from dgl.nn import SparseBiasedMultiheadAttention
+
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:290]: Maybe first declare `num_nodes = 100` and `num_edges = 80` to make the code more readable.

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 11]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        attn_drop=0.
+    ):
+        super().__init__()
+        self.feat_size = feat_size
+        self.num_heads = num_heads
+        self.head_dim = feat_size // num_heads
+        assert (
+            self.head_dim * num_heads == feat_size
+        ), "feat_size must be divisible by num_heads"
+        self.scaling = self.head_dim**-0.5
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:311]:
```suggestion
        self.scaling = self.head_dim ** -0.5

```

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 12]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        self.attn_bias_type = attn_bias_type
+
+        self.q_proj = nn.Linear(feat_size, feat_size, bias=bias)
+        self.k_proj = nn.Linear(feat_size, feat_size, bias=bias)
+        self.v_proj = nn.Linear(feat_size, feat_size, bias=bias)
+        self.out_proj = nn.Linear(feat_size, feat_size, bias=bias)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:317]: Not used

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 13]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        Number of attention heads, by which attr:`feat_size` is divisible.
+    bias : bool, optional
+        If True, it uses bias for linear projection. Default: True.
+    attn_bias_type : str, optional
+        The type of attention bias used for modifying attention. Selected from
+        'add' or 'mul'. Default: 'add'.
+
+        * 'add' is for additive attention bias.
+        * 'mul' is for multiplicative attention bias.
+    attn_drop : float, optional
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:282]: Perhaps directly call it `dropout`

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 14]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+
+        self.dropout = nn.Dropout(p=attn_drop)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Reset parameters of projection matrices, the same settings as that
+        in Graphormer.
+        """
+        nn.init.xavier_uniform_(self.q_proj.weight, gain=2**-0.5)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:327]: First declare `gain = 2 ** -0.5`, then reuse it.

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 15]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        nn.init.xavier_uniform_(self.k_proj.weight, gain=2**-0.5)
+        nn.init.xavier_uniform_(self.v_proj.weight, gain=2**-0.5)
+
+        nn.init.xavier_uniform_(self.out_proj.weight)
+        if self.out_proj.bias is not None:
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:332]: Seems to be that this should hold for the rest linear layers as well

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 16]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+            nn.init.constant_(self.out_proj.bias, 0.0)
+
+    def forward(self, g, ndata, attn_bias=None):
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:335]: Why not use `feat` or `nfeat` as in most existing NN modules? Same for `BiasedMultiheadAttention`

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 17]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+            The input graph of N nodes and E edges.
+        ndata : torch.Tensor
+            A 2D input tensor of node features. Shape: (N, :attr:`feat_size`)
+        attn_bias : torch.Tensor, optional
+            The attention bias used for attention modification.
+            Shape: (E, :attr:`num_heads`)
+
+        Returns
+        -------
+        y : torch.Tensor
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:350]:
```suggestion
        torch.Tensor
```

Same for `BiasedMultiheadAttention`

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 18]
```diff
@@ -216,6 +240,138 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+            The output tensor. Shape: (N, :attr:`feat_size`)
+        """
+        N = g.num_nodes()
+        with g.local_scope():
+            # linear projection
+            g.ndata['q_h'] = self.q_proj(ndata).reshape(
+                N, self.num_heads, self.head_dim
+            ) / self.scaling
+            g.ndata['k_h'] = self.k_proj(ndata).reshape(
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:359]: Not used at all

File: tests/pytorch/test_nn.py [hunk 19]
```diff
@@ -1801,3 +1801,25 @@ def test_BiasedMultiheadAttention(feat_size, num_heads, bias, attn_bias_type, at
     out = net(ndata, attn_bias, attn_mask)
 
     assert out.shape == (16, 100, feat_size)
+
+@pytest.mark.parametrize('feat_size', [128, 512])
+@pytest.mark.parametrize('num_heads', [8, 16])
+@pytest.mark.parametrize('bias', [True, False])
+@pytest.mark.parametrize('attn_bias_type', ['add', 'mul'])
+@pytest.mark.parametrize('attn_drop', [0.1, 0.])
```
Comment (mufeili) [inline tests/pytorch/test_nn.py:1809]: I think we can use a single value for all the options other than `attn_bias_type` to save test resources. Otherwise, this induces 2^5 tests.

**Review by mufeili — CHANGES_REQUESTED** [review]
Done a pass

File: python/dgl/nn/pytorch/graph_transformer.py (hunk 6 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/graph_transformer.py]: A Generalization of Transformer Networks to Graphs. In this paper, by default, they disable the bias in projection layers.
